sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,235 @@
+    X_test_decr = np.copy(X_test)
+    X_test_incr[:, 0] += 10
+    X_test_decr[:, 1] += 10
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = 1
+    monotonic_cst[1] = -1
+
+    for name, TreeRegressor in REG_TREES.items():
+        if depth_first:
+            est = TreeRegressor(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeRegressor(max_depth=None, monotonic_cst=monotonic_cst,
+                                max_leaf_nodes=n_samples_train)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": seed})
+        est.fit(X_train, y_train)
+        y = est.predict(X_test)
+        # increasing constraint
+        y_incr = est.predict(X_test_incr)
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict(X_test_decr)
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+    for name, TreeClassifier in CLF_TREES.items():
+        if depth_first:
+            est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst,
+                                 max_leaf_nodes=n_samples_train)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": seed})
+        est.fit(X_train, y_train)
+        y = est.predict_proba(X_test)[:, 0]
+
+        # increasing constraint
+        y_incr = est.predict_proba(X_test_incr)[:, 0]
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict_proba(X_test_decr)[:, 0]
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+
+def test_multiclass_raises():
+    X, y = datasets.make_hastie_10_2(n_samples=100, random_state=0)
+    y[0] = 0
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = -1
+    monotonic_cst[1] = 1
+    for name, TreeClassifier in CLF_TREES.items():
+        est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst,
+                             random_state=0)
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": 0})
+
+        with pytest.raises(ValueError):
+            est.fit(X, y)
+
+
+def assert_1d_reg_tree_children_monotonic_bounded(tree_, monotonic_sign):
+    # Flip values to always check for increasing constraint
+    values = monotonic_sign * tree_.value
+    for i in range(tree_.node_count):
+        if tree_.children_left[i] > i and tree_.children_right[i] > i:
+            # Check monotonicity
+            i_left = tree_.children_left[i]
+            i_right = tree_.children_right[i]
+            assert(float(values[i_left]) <= float(values[i_right]))
+            val_middle = float(values[i])
+            # Check bounds
+            if tree_.feature[i_left] >= 0:
```

Comment:
How can this not be the case? To filter out leaf nodes? Maybe a comment would make the code easier to follow.

Reply:
Indeed, the intention is just to filter out lead nodes. Adding a comment accordingly !